refactor webhook admission

hzxuzhonghu opened the pull request:
**What this PR does / why we need it**:

depends on #56478
refactor mutating and validating webhook admission plugins
1. deduplicate mutating and validating webhook code

TODO: 

- [x] ~refactor unit tests~

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

hzxuzhonghu committed:
admission registration use shared informer instead of poll

hzxuzhonghu committed:
run update bazel staging-dep

hzxuzhonghu committed:
Update bazel

hzxuzhonghu committed:
add wait ready for mutating/validating webhook configuration

sttts committed:
admission/webhook: refactor to webhook = generic-webhook + source + dispatcher

hzxuzhonghu committed:
admission/webhook: unify mutatuing validating test cases

hzxuzhonghu commented:
/assign @sttts @deads2k

hzxuzhonghu commented:
/test pull-kubernetes-bazel-test

hzxuzhonghu commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -0,0 +1,199 @@
+		Handler:       handler,
+		sourceFactory: sourceFactory,
+		clientManager: cm,
+		dispatcher:    dispatcherFactory(&cm),
```
the e2e failed cases related to this. Because the clientManager and the cm have not same address. So make clientManager as pointer may solve this.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/configuration/mutating_webhook_manager_test.go:
```diff
@@ -85,6 +88,26 @@ func (f *fakeMutatingWebhookConfigLister) Get(name string) (*v1beta1.MutatingWeb
 	panic("unsupported")
 }
 
+func newFakeMutatingWebhookConfigurationManager(f admissionregistrationinformers.MutatingWebhookConfigurationInformer) generic.Source {
```
these really look so similar. Any reason not to unify the tests? Does it work?

hzxuzhonghu replied:
Actually, I have not found a nice way to do unify.

sttts commented:
Can you squash the generated code commits into one?

sttts commented:
Can you link the base PR with the poll-informer change? We should get that in first.

hzxuzhonghu commented:
> Can you link the base PR with the poll-informer change? We should get that in first.

@sttts I want to close the previous one . Because it has conflicts and based on that this also have conflicts.

hzxuzhonghu commented:
/retest

sttts commented:
> I want to close the previous one . Because it has conflicts and based on that this also have conflicts.

Just push the changes from this PR into the old branch. I would like to avoid mixing PRs with different purposes.

hzxuzhonghu commented:
ok

hzxuzhonghu commented:
@sttts PTAL about the test unify.

sttts commented on staging/src/k8s.io/apiserver/pkg/admission/plugin/webhook/generic/webhook.go:
```diff
@@ -197,3 +199,33 @@ func (a *Webhook) Dispatch(attr admission.Attributes) error {
 }
+
+// GetConvertor returns the versioned convertor.
+func (a *Webhook) GetConvertor() *versioned.Convertor {
```
these getter and setting seem to be strange. Are they to inject mock objects? Maybe we need a way to inject them on construction. Having these getter and setters is not very Go idiomatic IMO.

hzxuzhonghu replied:
Yes, most are for mock objects, except this Convertor. And we can not inject them on construction, because this generic webhook is not called directly by test case, but by mutating and validating plugins. Or we can expose these fields。

sttts committed:
admission/webhook: fix panic from empty response in mutating webhooks

hzxuzhonghu commented:
/retest

sttts committed:
admission/webhook: remove broken VersionedAttributes override abstraction

This overriding had no effect. The versioned.Attributes were never used as admission.Attributes. Better
make the versioned objects explicit than hiding them under a wrong abstraction.

sttts committed:
admission/webhook: remove wrapping of scheme.Convert

sttts committed:
Update bazel

sttts committed:
admission/webhook: internalize conversion package

sttts committed:
Update bazel

hzxuzhonghu committed:
Merge pull request #7 from sttts/sttts-internalize-conversion

Make convertor internal to generic webhook

hzxuzhonghu commented:
close this one.
